packages/flutter_tools/lib/src/commands/update_packages.dart

```diff
@@ -53,39 +53,88 @@ class UpdatePackagesCommand extends FlutterCommand {
+            if (dependency.kind != previous.kind || dependency.lockTarget != previous.lockTarget) {
+              throw 'Inconsistent requirements around ${dependency.name}; '
+                    'saw ${dependency.kind} (${dependency.lockTarget}) in "${dependency.sourcePath}" '
+                    'and ${previous.kind} (${previous.lockTarget}) in "${previous.sourcePath}".';
+            }
           }
+          // Remember this dependency by name so we can look it up again.
           dependencies[dependency.name] = dependency;
+          // Normal dependencies are those we get from pub. The others we
+          // already implicitly pin since we pull down one version of the
+          // Flutter and Dart SDKs, so we track which those are here so that we
+          // can omit them from our list of pinned dependencies later.
           if (dependency.kind != DependencyKind.normal)
             specialDependencies.add(dependency.name);
         }
       }
 
-      final PubDependencyTree tree = new PubDependencyTree();
+      // Now that we have all the dependencies we explicitly care about, we are
+      // going to create a fake package and then run "pub upgrade" on it.
+      final PubDependencyTree tree = new PubDependencyTree(); // object to collect results
       final Directory temporaryDirectory = fs.systemTempDirectory.createTempSync('flutter_update_packages_');
       try {
         final File fakePackage = _pubspecFor(temporaryDirectory);
         fakePackage.createSync();
         fakePackage.writeAsStringSync(_generateFakePubspec(dependencies.values));
+        // First we run "pub upgrade" on this generated package:
         await pubGet(directory: temporaryDirectory.path, upgrade: true, checkLastModified: false);
-        await pub(<String>['deps', '--style=compact'], directory: temporaryDirectory.path, filter: tree.fill, retry: true);
+        // Then we run "pub deps --style=compact" on the result. We pipe all the
+        // output to tree.fill(), which parses it so that it can create a graph
+        // of all the dependencies so that we can figure out the transitive
+        // dependencies later. It also remembers which version was selected for
+        // each package.
+        await pub(
+          <String>['deps', '--style=compact'],
+          directory: temporaryDirectory.path,
+          filter: tree.fill,
+          retry: false, // errors here are usually fatal since we're not hitting the network
+        );
       } finally {
         temporaryDirectory.deleteSync(recursive: true);
       }
 
+      // Now that we have collected all the data, we can apply our dependency
+      // versions to each pubspec.yaml that we collected. This mutates the
+      // pubspec.yaml files.
+      //
+      // The specialDependencies argument is the set of package names to not
```

Comment:
Maybe call them "unpinned" rather than "special"?

Reply:
Well they are pinned, just not by a version number. I couldn't figure out a better term...

Reply:
Unversioned? Doesn't matter. The docs explain it well.